[Educational Notes] Start adding educational notes for data-race safety. (#79509)
swiftlang/swift: userdocs/diagnostics/sendable-closure-captures.md

# Captures in a `@Sendable` closure

> `@Sendable` closures can be called multiple times concurrently, so any captured values must also be safe to access concurrently. The compiler will prevent mutable captures in a `@Sendable` closure under complete concurrency checking.

Reviewer: ```suggestion
Marking a closure as `@Sendable` tells the compiler the closure can be called multiple times concurrently, so any captured values must be safe to access concurrently. The compiler prevents capturing mutable values in a `@Sendable` closure under complete concurrency checking.
```

For example:

```swift
func callConcurrently(
  _ closure: @escaping @Sendable () -> Void
) { ... }

func capture() {
  var result = 0
  result += 1
  
  callConcurrently {
    print(result)
  }
}
```

Building with complete concurrency checking will diagnose the capture of `result` in a `@Sendable` closure:

```
|   callConcurrently {
|     print(result)
|           `- error: reference to captured var 'result' in concurrently-executing code
|   }
| }
```

> Because closure is `@Sendable`, the implementation of `callConcurrently` is allowed to call `closure` multiple times concurrently, e.g. using a task group:

Reviewer: ```suggestion
Because the closure is marked `@Sendable`, the compiler allows the implementation of `callConcurrently` to call `closure` multiple times concurrently, for example using a task group:
```

```swift
func callConcurrently(
  _ closure: @escaping @Sendable () -> Void
) {
  Task {
    await withDiscardingTaskGroup { group in
      for _ in 0..<10 {
        group.addTask {
          closure()
        }
      }
    }
  }
}
```

> If the type of the capture is `Sendable` and the closure only needs the value of the variable at the point of capture, you can resolve the error using capture by value in a capture list:

Reviewer 2: ```suggestion
If the type of the capture is `Sendable` and the closure only needs the value of the variable at the point of capture, you can resolve the error by explicitly capturing this variable in the closure's capture list:
```

```swift
func capture() {
  var result = 0
  result += 1
  
  callConcurrently { [result] in
    print(result)
  }
}
```

This strategy does not apply to captures with non-`Sendable` type. Consider the following example:

```swift
class MyModel {
  func log() { ... }
}

func capture(model: MyModel) async {
  callConcurrently {
    model.log()
  }
}
```

> Building with complete concurrency checking will diagnose the capture of `model` in a `@Sendable` closure:

Reviewer: ```suggestion
Building with complete concurrency checking diagnoses the capture of `model` in a `@Sendable` closure:
```

```
| func capture(model: MyModel) async {
|   callConcurrently {
|     model.log()
|     `- error: capture of 'model' with non-sendable type 'MyModel' in a '@Sendable' closure
|   }
| }
```

> If a type with mutable state can be referenced concurrently but all access to mutable state happens on the main actor, the best way to model that is isolating the type to a global actor and marking the methods that don't access mutable state with `nonisolated`:

Reviewer: ```suggestion
If a type with mutable state can be referenced concurrently but all access to mutable state happens on the main actor, isolate the type to the main actor and mark the methods that don't access mutable state with `nonisolated`:
```

Reviewer: ^ I've seen conflicting advice on this point, and understand that some developers that see "mixed" isolation within a type is a "bad code smell". There may be times when it can't be avoided, but I wanted to mention that since this is going into a compiler-advising kind of output.

Author: I definitely don't agree that using `nonisolated` inside an actor isolated type is a code smell. I think this is a great pattern - there will be properties and methods on an isolated type that are totally safe to call from any concurrency domain, e.g. constants or methods that just don't access any mutable state.

> If a type with mutable state can be referenced concurrently but all access to mutable state happens on the main actor, the best way to model that is isolating the type to a global actor and marking the methods that don't access mutable state with `nonisolated`:

Reviewer 2: ```suggestion
If a type with mutable state can be referenced concurrently but all accesses to mutable state happen on the main actor, the best way to model that is to isolate such type to a global actor and mark its methods that don't access mutable state with `nonisolated`:
```
